### ISSUE TITLE
WS2812B OpenELEC 6.03 RPi2 Guide

### ISSUE THREAD

**Issue by r3dr4bbit** [tvdzwan/hyperion#525]
No an issue, but since I spent two days getting this two work and everyone else is struggling as well, here is the how to:

1. Download the latest image from OpenELEC
2. In config.txt add the following to the: "dtoverlay=pwm"
3. Install Hyperion using the normal script as found in the wiki
4. In the following post, there is a link to Zappatrons' Google Drive where a recompiled hyperiond is located.  Download it and paste it in /storage/hyperiono/bin.
https://github.com/tvdzwan/hyperion/issues/6#issuecomment-77109750
5. Remember to upload configuration file
6. Also remember to change /opt to /storage in the config

**Comment by polykoma** [comment]
Thanks for sharing!
Can someone confirm this workaround?

**Comment by NicoHood** [comment]
The wiki is a good place to add this information:
https://github.com/tvdzwan/hyperion/wiki

**Comment by r3dr4bbit** [comment]
@polykoma I just did, busy performing it again...

**Comment by brindosch** [comment]
Just to clarify. If you use the official source and add the ws2812b_pwm flag for compiling this just works for Rpi 1 not for 2?

**Comment by r3dr4bbit** [comment]
nope.  The official source was forked to Zapatron's repo, who fixed the memory values.  Hence the download of the compiled download.

You are also more than welcome to clone Zapatron's repo and build it there with the WS2812b_pwm flag enabed.

**Comment by brindosch** [comment]
Is there a possibility to merge that to one?
Or is his source now just Rpi2 capable?

**Comment by r3dr4bbit** [comment]
Check with the two authors @tvdzwan and @Zappatron.

**Comment by NicoHood** [comment]
We could just compile 2 versions or make a better fix. No need to download some file from a google drive link. A Proper rpi2/3 fix should be made anyways.

**Comment by Zappatron** [comment]
The changes I made are just a quick and dirty hack, to be merged it would need to be done in a much nicer way, autodetecting the memory addresses required based on the device it's running on - or at least a new device driver specifically for the pi2/3.

**Comment by brindosch** [comment]
@Zappatron
If you are motivated, just do it :+1: 
All users would be very thankful for this.

**Comment by kiryl85** [comment]
Hi @Zappatron.
May I ask you something? I'm using OSMC on RPi2 and would like to check ws2812b leds on the newest hyperion connected directly to GPIO. At the moment I'm using them successfully through Arduino.
So, in order to have ws2812b working I should replace original  libsrc/leddevice/LedDeviceWS2812b.cpp file with your one and set ENABLE_WS2812BPWM flag during compilation, am I right?

**Comment by r3dr4bbit** [comment]
Hi @kiryl85 
Pretty much yes.  Also remember to place the pwm_overlay.dtb boot partition and also register it using dtoverlay=pwm.

**Comment by Zappatron** [comment]
@kiryl85 Unfortunately, as it stands, the 2812 pwm led driver doesn't work on OSMC since the November 2015 release. I haven't done any investigation as to why.

**Comment by r3dr4bbit** [comment]
@Zappatron I suspect its worth looking into the overlay's as well?  No recompile required.  I will see if I can get something running later today.

**Comment by Zappatron** [comment]
Oh, so it's possible to get it working on OSMC > 201511 now using an overlay? I did try a few things last year but haven't looked in to it since.

I'll take a look at the weekend if I get the time.

**Comment by r3dr4bbit** [comment]
Well, as far as I know in parallel with OSCM going to 201511 build, OpenELEC went to 5.0.6, in which case Hyperion also stopped working (for me at least).  This is about the same time they bought Device Tree Overlays into the RPi kernel which I suspect is the main suspect in all the dilemmas.

Yesterday I struggled long and hard to get Hyperion working.  I finally got it working on OpenELEC 5.0.1, updated and then it stopped working.  Once I added the pwm overlay, it worked again.

**Comment by Zappatron** [comment]
Sounds promising. I'll confirm once I've tested.

**Comment by brindosch** [comment]
If we could merge all to one image and add a new wiki article,
i would write a new install script that do all steps for you (Install_hyperion_PWMONLY.sh

**Comment by penfold42** [comment]
Maybe the quickest way to merge is to add another led device type ? 

W2812-PWM-rpi01 and W2812-PWM-rpi23 ?

**Comment by redPanther** [comment]
I've took a quick look into the code.
Merging shouldn't be hard.
The quickest way is a compile option. I'll investigate this issue tomorow evening.

Cheers, rp

**Comment by Zappatron** [comment]
@r3dr4bbit Unfortunately enabling the pwm overlay still doesn't seem to work in the latest OSMC

**Comment by redPanther** [comment]
@Zappatron 
>  autodetecting the memory addresses required based on the device it's running on

atm I have no time, but I have a solution for detection if we are on rpi 1 or higher.
This solution is not the best, because the revision field of /proc/cpuinfo is more reliable.
here is an implemention for getting full info about rpi:
https://github.com/AndrewFromMelbourne/raspberry_pi_revision

This is my solution - should be enough for deciding wich base adress to use.

```
#include <fstream>
#include <sstream>
#include <string>
#include <iostream>

int detectRaspberryPiVersion()
{
        std::ifstream infile("/proc/cpuinfo");
        std::string line;
        while (std::getline(infile, line))
        {
                if (line.compare(0,8,"Hardware") == 0 && line.compare(line.size()-7,7,"BCM2708") == 0) {
                        return 1;
                }
                if (line.compare(0,8,"Hardware") == 0 && line.compare(line.size()-7,7,"BCM2709") == 0) {
                        return 2;
                }
        }
        return 0;
}


int main()
{
        int piVersion = detectRaspberryPiVersion();
        if ( piVersion )    std::cout << "found rpi " << piVersion << std::endl;
        else                  std::cout << "unknown pi version" << std::endl;
        return 0;
}
```

**Comment by brindosch** [comment]
Side note: RPi3 will be BCM2710 (not now, but the files are there. Maybe a switch to armv8 in future)

**Comment by brindosch** [comment]
To provide better support for pwm users. (At least through the latest commit!)
I extended the install script. Testers are welcome.

Feature: Adds the dtoverlay=pwm automatically if not set (all rpi system/distros)

The usage is easy, just start the script with the pwm argument like shown bottom.

```
curl -# -k -L --output remove_hyperion.sh --get https://raw.github.com/brindosch/hyperionbeta/master/bin/remove_hyperion.sh
sudo sh ./remove_hyperion.sh WS281X
```
The images are even with hyperion master, so no downgrade for you

**Comment by kiryl85** [comment]
hi @brindosch. May I know what is the usage for PWM variable in your script? I see it's set to "1" after using WS281X argument but that's all. PWM is not used anymore.

**Comment by brindosch** [comment]
You are so right, i just pushed this very fast and never tested it. I was sure all is in place :)
Now it should work :)

**Comment by denn64** [comment]
Tell me, do I need to connect to my raspberries pi2 gipo18? 
I understand that the tape ws2812b work directly with raspberries 2? has more detailed instructions ? I do not quite understand everything because of Russia . sorry for my English

**Comment by penfold42** [comment]
Yes - you use gpio18 for the data connection. You should use a level shifter if your strip is powered by 5V.

Only the "type" : "ws281x" supports rpi1,2 and 3. The original ws2812b only supports rpi1

**Comment by denn64** [comment]
and who can provide a link to an image with a ready-made solution raspberry pi2 ws2812b. I did not poluchaetsya tape does not burn .

**Comment by brindosch** [comment]
@Zappatron 
Could you code a fix on @redPanther  idea?
I am sure the users would love it :)

**Comment by kiryl85** [comment]
hi @penfold42. My current setup is based on arduino via USB and it works great. But I would like to check connection directly via GPIO.
So on RPi2 it should work as long as mentioned level shifter for data is used, correct? Could you please provide me with a link to a correct level shifter?
 In addition, can 5V and GND for this level shifter be taken from GPIO pins?

**Comment by penfold42** [comment]
I've never used a level shifter personally - my testing with a short strip was ok.
Google found this: http://happyinmotion.com/?p=1247

The 5v for the level shifter is fine from the gpio header but don't power the led strip from the gpio.

**Comment by denn64** [comment]
Tell me, is it possible to run on ws2812b raspberry pi2? or necessarily need arduino? I can not understand. there are detailed instructions ?

**Comment by kiryl85** [comment]
Thank you @penfold42. I will take a closer look on it. Right now I'm wondering if I can test connection straight from RPi2 without level shifter   to see the results. Can it damage my LED strip? 

Also, @brindosch, I'm sorry for another uncomfortable question but which version is installed by your script? The one without PWM functionality, am I right? 
Additional ws.. parameter is used only for adding PWM overlay, is that correct?

**Comment by penfold42** [comment]
@kiryl85 if you wire it correctly it is safe.
I ammended the wiki https://github.com/tvdzwan/hyperion/wiki/WS2812b-PWM-led-devices to include discussion of the new driver and there is an existing diagram there.

**Comment by brindosch** [comment]
@kiryl85 
With both repos (my and hyperion) you will get a pwm enabled version.
As you saw above, there is no support for 2812B on Pi2/Pi3 as long no one implements it :)

At the moment it is just to add the dtoverlay=pwm to config.txt and ensure the user is correctly set on osmc. If we need more, i will add it.

**Comment by Culloden-Spectre** [comment]
@brindosch 
Mate, thank you so much for your script!
I have tested your scripts using OpenELEC (latest version as of yesterday) and could not get it to work, even took out my old CRO and there was no activity on the pins. not sure what has gone wrong on my end but i spent a couple of days testing and could not get it working for me.

Then decided to run your script on a fresh install of OSMC, changed the config.txt to "ws281x", and my CRO showed a PWM signal, plugged in my lights and everything worked flawlessly.

**Comment by penfold42** [comment]
I'm struggling to get mine working on OSMC.
Can you please send me :
What model pi ?
uname -a
cat /boot/config.txt
lsmod

**Comment by Culloden-Spectre** [comment]
[RPi3Config.txt](https://github.com/tvdzwan/hyperion/files/178993/RPi3Config.txt)

@penfold42 hope this helps

**Comment by penfold42** [comment]
Thanks so much - will divide and conquer the differences !

**Comment by brindosch** [comment]
The new installer should do all necessary steps now. Just read at the OpenELEC/OSMC section!
If you have problems, please try first a update with the install script.
/closed
